pkg/synchronizer/start_gate.go

```diff
@@ -0,0 +1,161 @@
+// considered "live". Callers push packets into the gate until it returns a
+// slice of packets that can be used to initialize downstream synchronisation.
+// The second return value exposes how many packets were discarded while the
+// gate was waiting for stability, and the third return value indicates whether
+// the gate has finished its job.
+type startGate interface {
+	Push(pkt jitter.ExtPacket) ([]jitter.ExtPacket, int, bool)
+}
+
+// burstEstimatorGate implements startGate using the burst-estimation logic.
+// It buffers packets until their arrival cadence matches
+// the RTP timestamp spacing well enough to assume we are caught up with the
+// realtime stream. The gate can be configured with different tolerances for
+// audio and video since video keyframes may span hundreds of packets.
+type burstEstimatorGate struct {
+	clockRate   uint32
+	maxSkew     time.Duration
+	minArrival  time.Duration
+	scoreTarget int
+
+	score       int
+	lastTS      uint32
+	lastArrival time.Time
+	hasLast     bool
+	done        bool
+	buffer      []jitter.ExtPacket
+	maxBuffer   int
+}
+
+func newStartGate(clockRate uint32, kind webrtc.RTPCodecType) startGate {
+	be := &burstEstimatorGate{
+		clockRate:   clockRate,
+		maxSkew:     20 * time.Millisecond,
+		minArrival:  2 * time.Millisecond,
+		scoreTarget: 5,
+		maxBuffer:   1000, // high bitrate key frames can span hundreds of packets
+	}
+
+	if kind == webrtc.RTPCodecTypeAudio {
+		be.maxSkew = 8 * time.Millisecond
+		be.maxBuffer = 200
+	}
+
+	return be
+}
+
+// Push feeds one packet into the gate. Once pacing stabilizes it returns the
+// buffered packets that should be used to initialize the track synchronizer,
+// along with the number of packets dropped while waiting and a done flag.
+func (b *burstEstimatorGate) Push(pkt jitter.ExtPacket) ([]jitter.ExtPacket, int, bool) {
+	if b.done {
+		return nil, 0, true
+	}
+
+	if len(pkt.Payload) == 0 {
+		return nil, 0, false
+	}
+
+	if !b.hasLast {
+		b.lastTS = pkt.Timestamp
+		b.lastArrival = pkt.ReceivedAt
+		b.hasLast = true
+		b.buffer = append(b.buffer[:0], pkt)
+		return nil, 0, false
+	}
+
+	signedTsDelta := int32(pkt.Timestamp - b.lastTS)
+	arrivalDelta := pkt.ReceivedAt.Sub(b.lastArrival)
+
+	if signedTsDelta < 0 {
+		// ignore out-of-order packets while continuing to wait for stability
+		return nil, 0, false
+	}
+
+	if signedTsDelta == 0 {
+		// multiple packets with the same timestamp (e.g. key frame)
+		b.buffer = append(b.buffer, pkt)
+		if dropped := b.enforceWindow(); dropped > 0 {
+			return nil, dropped, false
+		}
+		return nil, 0, false
+	}
+
+	b.lastTS = pkt.Timestamp
+	b.lastArrival = pkt.ReceivedAt
+
+	tsDuration := b.timestampToDuration(uint32(signedTsDelta))
+	if tsDuration <= 0 {
+		dropped := len(b.buffer)
+		b.buffer = b.buffer[:0]
+		b.score = 0
+		b.buffer = append(b.buffer, pkt)
+		return nil, dropped, false
+	}
+
+	if arrivalDelta < b.minArrival {
+		dropped := len(b.buffer)
```

Comment:
For this kind of code block, I usually make a small function variable and call it to avoid repeated code like this. Works better if I have to change something later. Just have to change one place.